### DISCUSSION TITLE
Coding standards, style guides and tools, General Best Practices for good code

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#277] [Ideas]
**General Best Practices for coding, js and node**:
https://www.lexjansen.com/pharmasug/2011/TT/PharmaSUG-2011-TT05.pdf
https://opensource.com/article/17/5/30-best-practices-software-development-and-testing
https://levelup.gitconnected.com/javascript-best-practices-for-writing-more-robust-code-finding-items-49b4ebe97d04
https://medium.com/dataseries/javascript-best-practices-robust-code-ab8e8a3a6524
https://www.perfomatix.com/nodejs-coding-standards-and-best-practices-node-js-development-company/
https://blog.risingstack.com/node-js-best-practices/
https://javascript.info/

**Style Guides**:
We have quite a few style guides:

We have 5 mentioned here: https://codeburst.io/5-javascript-style-guides-including-airbnb-github-google-88cbc6b2b7aa
https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/JavaScript
https://www.w3schools.com/js/js_conventions.asp
https://google.github.io/styleguide/jsguide.html
A styleguide apparently used by Node community: https://github.com/felixge/node-style-guide
Also this for Node: https://medium.com/swlh/node-js-coding-style-guidelines-74a20d00c40b
One more: https://javascript.info/coding-style

We need to pick one style guide for this project and adhere to it strictly. We should pick the one which has the most tooling i.e. tools which can detect deviations and suggest corrections (even autocorrect if possible.) These tools should work well with Open Source IDEs - especially Eclipse and VSCode for sure and Atom too if possible. Check if these tools can be integrated to the popular proprietary IDE Webstorm too.

**Comment by kcramakrishna** [comment]
- Another thing that's great to include in comments are URLs to wherever you found a solution to your coding problem online.
- Avoid ninja-code at all costs: https://javascript.info/ninja-code

Linters:

CSS:
https://cssdeck.com/blog/best-active-css-lint-tools/
Basically boils down to http://csslint.net/ vs. https://stylelint.io/

stylelint:
https://css-tricks.com/stylelint/
https://github.com/morishitter/stylefmt

TBD: Pick the right css linter

Javascript:
https://www.codeguru.com/tools/the-top-five-javascript-linting-tools/

Basically boils down to ESLint (https://eslint.org/) vs. standardJS (https://standardjs.com/)
https://www.npmtrends.com/eslint-vs-standardjs
There is a way to use eslint with standard: https://dev.to/itmayziii/typescript-eslint-and-standardjs-5hmd

standardjs uses eslint underneath. Looks like standardjs makes our life easier. It has pre-commit hooks too.
https://github.com/standard/standard
Standardjs usage: https://npm-stat.com/charts.html?package=standard&author=&from=&to=
Has a ton of stars on github

More research and discussion is needed here. standardjs seems to be the front-runner
TBD: Pick the right javascript linter

**Reply by kcramakrishna** [reply]
Do we need prettier on top of standardjs? 
https://prettier.io/
https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

**Comment by yureshwar** [comment]
Implemented eslint based configuration for source code analysis via git hooks. Followed the tutorial from the below given link

https://www.antstack.io/blog/adding-git-hooks-to-your-project/

**Comment by kcramakrishna** [comment]
There seems to be good libraries for validations
https://dev.to/samchon/typescript-json-is-10-1000x-times-faster-than-zod-and-io-ts-8n6
https://dev.to/captainyossarian/typescript-static-validation-4oln

I think its a good idea to add validations to all functions/methods for input params and also verify just before returning anything.

**Comment by kcramakrishna** [comment]
Found this new faster json stringify library - https://dev.to/samchon/i-made-10x-faster-jsonstringify-functions-even-type-safe-2eme 
We should consider using it here.

**Reply by kcramakrishna** [reply]
Here is the research from github - https://github.com/samchon/typescript-json
1,400 stars, 21 contributors
Insights not very healthy - https://github.com/samchon/typescript-json/graphs/contributors 

Process involved:
1. Fork the domjson library
2. Replace stringify all over 
3. Use the forked domjson in our app.
Risk: Upgrade to new version of domJSON will be problematic. 
Observation: donJSON is unmaintained and the last release was in 2015. https://github.com/azaslavsky/domJSON/releases
This means that upgrade is not likely and we not face upgrade problems i.e. merging into our code base. 

Benefit to our project:
We have have 2 places where we are using stringify - 
1. When we are recording the node.
2. When we are compreing nodes during playback. i.e. We are comparing ALL nodes in that page with the 'recorded' node. This can be hundreds (sometimes thousands) of comparisons during playback of every single node. Hence any speed increase here will be very useful to our project.

Given the promise of performance improvement, despite the extra effort of fork+change process, we should definitely use this in our project.
We will not use any of the other forks since we do not know what changes have been made. Will use original sourcecode.

Course of action:
Stage 1: We can fork this project and make performance change and optionally call it 'fastDomJSON. Do this immediately as this is simpler.
https://github.com/Digital-Assistant/Digital_Assistant_Client/issues/362 

Future and unscheduled: 
Stage 2: The project seems to be abandoned. Consider other libraries.
Maybe fork this to 'newdomjson' modernise whole source-code to typescript and upgrade performance.

Alternatives: to domJSON:
https://www.npmjs.com/package/dom-to-json - https://github.com/sumn2u/dom-to-json
https://rapidjson.org/index.html - This looks very promising for json comparison. **We should research this more**. https://github.com/Digital-Assistant/Digital_Assistant_Client/issues/363
